tools/crm_resource.c

```diff
@@ -965,7 +965,70 @@ main(int argc, char **argv)
         rc = pcmk_ok;
 
     } else if (rsc_cmd == 'e') {
+        xmlXPathObject *xpathObj = NULL;
+        xmlNode *cib_constraints = NULL;
+        crm_time_t *now = crm_time_new(NULL);
+        int i;
+
+        cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS,
+                                          data_set->input);
+        xpathObj = xpath_search(cib_constraints, "//" XML_CONS_TAG_RSC_LOCATION);
+
+        for (i = 0; i < numXpathResults(xpathObj); i++) {
+            xmlNode *constraint_node = getXpathResult(xpathObj, i);
+            xmlNode *date_expr_node = NULL;
+            crm_time_t *end = NULL;
+            char *xpath_string = NULL;
+
+            if (!crm_starts_with(ID(constraint_node), "cli-ban-") ||
+                !crm_starts_with(ID(constraint_node), "cli-prefer-")) {
+                continue;
+            }
+
+            /* Find the date_expression within this constraint. */
+            xpath_string = crm_strdup_printf("//date_expression[@id='%s-lifetime']",
+                                             ID(constraint_node));
+            date_expr_node = get_xpath_object(xpath_string, constraint_node, LOG_DEBUG);
+
+            if (date_expr_node == NULL) {
+                free(xpath_string);
+                continue;
+            }
+
+            /* And then finally, see if the date expression is expired.  If so,
+             * clear the constraint.
+             */
+            end = crm_time_new(crm_element_value(date_expr_node, "end"));
+
+            if (crm_time_compare(now, end) == 1) {
+                xmlNode *fragment = NULL;
+                xmlNode *location = NULL;
+
+                fragment = create_xml_node(NULL, XML_CIB_TAG_CONSTRAINTS);
+                location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
+                crm_xml_set_id(location, "%s", ID(constraint_node));
+                crm_log_xml_info(fragment, "Delete");
+
+                rc = cib_conn->cmds->remove(cib_conn, XML_CIB_TAG_CONSTRAINTS,
+                                            fragment, cib_options);
+                if (rc != pcmk_ok) {
+                    free_xml(fragment);
+                    crm_time_free(end);
+                    free(xpath_string);
+                    goto bail;
+                }
+
+                free_xml(fragment);
+            }
+
+            crm_time_free(end);
+            free(xpath_string);
+        }
+
+        freeXpathObject(xpathObj);
+
         rc = pcmk_ok;
+        crm_time_free(now);
```

Comment:
Let's functionize the new code and put it in crm_resource_ban.c. We have a bit of a bloat problem in crm_resource :)